### DISCUSSION TITLE
FDYQ (X35A/X50A) unable to control - protocol error

### DISCUSSION THREAD

**Discussion by howlzj** [revk/xESP32-Faikout#248] [General]
I have a FDYQ ducted system (AUS, 2010) which has an X35/X50A port not recognised by the Faikin.

It's generating these errors in MQTT as it cycles through the Tx/Rx combinations:

{"protocol":"X50A","badhead":true,"mismatch":true,"badrxlen":true,"data":"15EA"}
{"protocol":"X50A¬Tx","badhead":true,"mismatch":true,"badrxlen":true,"data":"15EA"}
{"protocol":"X50A¬Rx","badsum":"7A","data":"7505"}
{"protocol":"X50A¬Tx¬Rx","badsum":"F4","data":"75057505"}

I'll mention the X50A port on this model is a different size to others here. It's an 8-pin JST-EH connector (same as S21 but longer) - I had to modify the cable to connect it. The mainboard is identical to the one in this eBay listing: https://www.ebay.co.uk/itm/224203252109

What do I need to start debugging? It's a ducted unit and a _real_ pain to get in there.

**Comment by revk** [comment]
You can set `dump` to see more, but that is very odd. It is seeing some data though. I mean, silly one, try power cycling the whole lot just in case it has someone how got in a fix.

You can also disable the inverted tx and rx attempts and disable s21 in the settings to stop it trying anything else but the straight X50 protocol. (and then power cycling it all).

`15EA` rings a bell for some reason, but no idea why.

**Comment by howlzj** [comment]
No luck unfortunately. I also tried rewiring the harness just in case and power cycling a few times with the suggestions above.

Not sure if anyone has experience connecting the wrong X50A pins (if that's a possible cause). I made the assumption JST socket orientation (therefore the pinouts) would match between models, so my harness is connected like this:

---X50A--- (comparing to the eBay image linked above)
[----XY-Z] (where X=GPIO34, Y=GPIO48, Z=5V)

In your code, do you just pass along the raw message from the unit, or is 15EA a parsed response?

Whilst I am getting an Rx response, it's completely unresponsive to the Tx message:


Rx side:
- {"protocol":"X50A","dump":"15EA"}
Tx side:
- {"protocol":"X50A","dump":"06AA0701000146"}
- {"protocol":"X50A","cmd":"CB","payload":"0000"}
- {"protocol":"X50A","cmd":"CA","payload":"0000000000000000000000000000000000"}

**Reply by revk** [reply]
I think `06AA0701000146` is OK... Is this your own board?

`dump` is the raw sequence received.

It may be you don't have tx working somehow, and so are just seeing blind messages coming in.

If so, set GPIOs in config with or without a `-` prefix depending on whether you have inverting in place on them

**Reply by howlzj** [reply]
The Faikin is direct from the Amazon page and is reporting its ID as `F412FA97CEF8` in mqtt, so I'm not sure where `06AA0701000146` is coming from. Have also climbed up and replaced the harness (Aliexpress) with some known-good dupont cables _just in case_.

Inverting the GPIOs is definitely doing _something_.

When Rx _is_ inverted, I'll get:
Error: `{"protocol":"X50A","badhead":true,"mismatch":true,"badrxlen":true,"data":"15EA"}`
Dump: `{"protocol":"X50A","dump":"15EA"}` (sometimes 15EA15EA repeated

When Rx _is not_ inverted, I'll get:
Error: `{"protocol":"X50A","badsum":"7A","data":"7505"}` (sometimes F4 / 75057505)
Dump: `{"protocol":"X50A","dump":"7505"}`

Inverting Tx either way has no obvious effect.

**Reply by revk** [reply]
Err, `06AA0701000146` is data it has seen from the air-con. I think that is sensible.

So put GPIOs back with `-` prefix, as the Faikin does invert both.

And check your have the cables connected properly. It may be Tx is not getting through...

Do you have anything connected to pin 1 on the Faikin? My unit (X50A port) does not.

**Reply by howlzj** [reply]
Hmm. I can't get Loopback recognised either (stays flashing magenta) so this is potentially unrelated to my A/C!

I connected the Tx/Rx GPIO pairs and still getting `cannot-tx` (and weirdly that same `06AA0701000146`) even after toggling everything back to (AFAIK) defaults.

Is there a way to fully reset the settings on these, and I'll see if these phantom messages keep coming through? Looks like default OTA update script keeps all the settings.

**Reply by revk** [reply]
Yeh, loop back only works if there is a pull up on the Tx, so not surprised. Sorry.

Cannot tx is because it has not seen the sensible incoming messages, and so thinks off line.

`06AA0701000146` is definitely what we expect to see. length `06`, cmd `AA`, and not `00` meaning it is ready, so should be happy with that.

There is an MQTT command to factory reset, `command/Faikin/factory` with payload that starts with the MAC and `Faikin`.

Setting not to invert tx or rx and not to try s21 should mean it just talks X50, and that should work, but see what dump says in that mode, more than just one message, and maybe that will give some clues.

And nothing on pin 1, right?

**Reply by howlzj** [reply]
Nothing on pin 1. No effect if connected.

Pretty sure the pinout is correct - or at least it's the only pair that at least gives `badsum` rather than `timeout`.

Unfortunately I'm only seeing this very small number of messages, repeating every ~1 second depending if the GPIO is inverted or not (n.b. same with the auto invert turned off, below just for clarity).

Rx:
`{"protocol":"X50A","dump":"15EA"}` (sometimes returns `15EA15EA` so I guess an echo?)
`{"protocol":"X50A¬Tx","dump":"15EA"}`
`{"protocol":"X50A¬Rx","dump":"7505"}` (sometimes returns `75057505` so I guess an echo?)
`{"protocol":"X50A¬Tx¬Rx","dump":"7505"}`

Tx:
`{"protocol":"X50A","cmd":"AA","payload":"01"}` returns:
- `{"protocol":"X50A","dump":"06AA0701000146"}`
- `{"protocol":"X50A¬Tx","dump":"06AA0701000146"}`
- `{"protocol":"X50A¬Rx","dump":"06AA0701000146"}`
- `{"protocol":"X50A¬Tx¬Rx","dump":"06AA0701000146"}`

I have a regular ESP32 WROOM dev board with ESPHome installed - I believe it can sniff UART so will do some research what it might be able to sniff.

**Reply by revk** [reply]
OK this is confusing as you are not showing the topic on each message which tells if, for example, the `"dump"` is `tx` or `rx`.

i.e. I would expect to see `tx` and then `rx` and then `tx` and then `rx`.

I think you may this time be telling me all the `rx` in a row and then all the `tx` in a row.

So it may be the `06AA0701000146` is simply a dump of what we are sending, which is why it looks right. The correct reply looks very similar.

Also, it is doing the tx and rx invert attempts, there are settings, `noswaptx` and `noswaprx` which you probably should turn on to avoid some confusion, and `nos21` to stop it trying S21 as well.

Then you should see a clear `tx` dump like `06AA0701000146` and an `rx` dump of some sort after each.

The GPIO settings should be `tx` as `-48` and `rx` as `-34`, as that matches the hardware.

Once you have that, it may be clearer what is happening.

What is worrying is that those responses look a lot like wrong Baud rate, even what you might see if it was S21 not X50A even. But let's try one step at a time. Make sure the GPIOs are set as those, and the `swap` options are off.

**Reply by howlzj** [reply]
Ah - that's probably a function of me copying straight from MQTT Explorer (it copies the payload not the tx/rx topic).

I had a spare ESP32 and put ESPAltherma on it. Looks like Daikin Australia may have been a little naughty with their units here...

Looking at their [guide](https://raomin.github.io/ESPAltherma/), I discovered:
- my X50A looks just like the X10A port on their ROTEX brand units.
- ESPAltherma has a specific section 0x**15** 0x**EA** error (lol 15EA) - meaning it's specifically an _Altherma_ range unit with old 'S' protocol.
- Models on 'S' protocol will also not communicate if GND is not connected.

Conclusion: Daikin AU built this FDYQ model using older protocol 'S' _and_ threw in a mislabelled port just to confuse everyone.

Unfortunately the scope of ESPAltherma is read-only access, but I at least know the unit is responding!

`{
    "EV (pls)": 0,
    "Outdoor Fan (Upper)(rps)": 5,
    "Outdoor Fan (Lower)(rps)": 5,
    "INV Comp. Frequency(Hz)": "OFF",
    "Comp. Preheat": "ON",
    "52C Output": "ON",
    "20R (SV) Output": "OFF",
    "20S (4-way) Output": "OFF",
    "Crankcase Heater": "OFF",
    "Ener-Cut Output": "OFF",
    "Indoor Suction Air Temp.(C)": 24.5312,
    "Indoor Heat Exchanger Temp.(C)": 24.9375,
    "Outdoor air temp.(C)": 25.2695,
    "Outdoor heat exchanger temp.(C)": 0,
    "Discharge pipe temp.(C)": 48,
    "Fin Temp.(C)": 0,
    "Delta-Tr(deg)": 0,
    "R/C Setpoint(C)": 0,
    "Operation Mode": "Dry",
    "Error Code": " 0",
    "Thermo Off Error": " 0",
    "Warning Code": " 0",
    "Caution Code": " 0",
    "WifiRSSI": "-45dBm",
    "FreeMem": "229712"
}`

**Reply by revk** [reply]
OK so does setting S21 (`nox50a` set and `nos21` unset) work?

**Reply by howlzj** [reply]
No - it's definitely not working on S21 protocol either (even grounding the port). Sorry, should have been clearer - my unit may be from a range of Daikin that isn't (yet!) compatible with your design.

I stumbled on the ESPAltherma project - their scope is about the heating / hot water range of Daikin. They've concluded there are Daikins out there using a slightly different serial protocol (documented [here](https://github.com/raomin/ESPAltherma/blob/main/doc/Daikin%20S%20protocol.md)) which happened to work with my unit (they also use an ESP32 reporting to MQTT)

**Reply by revk** [reply]
Well, we may be able to do something with that!

Is that something you would like me to work on with you?

**Reply by howlzj** [reply]
Sure, we can give it a go - though I'm not sure if this bus is read-only or read-write.

The unit (fortunately) has very basic controls:

- Fan: High / Low
- Mode: Fan, Dry, Auto, Cool, Heat
- Temp: X-Y

I (think) I've located the relevant positions in the response. In ESPAltherma they've defined each reading to a string / position / lookup rule (some lookup rules apply maths). The relevant ones are below (removed ... descriptions that were garbage)

A/C On/Off - {0x53, 6, 200, 1, -1, "..."},

- ON: `0x53 0x00 0x00 0x03 0x01 0x00 0x00 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xa7` 
- OFF: `0x53 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xab` 

Fan Hi/Lo - {0x53, 2, 164, 1, -1, "..."},

- High: `0x53 0x00 0x00 0x03 0x01 0x00 0x00 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xa7` 
- Low: `0x53 0x00 0x00 0x02*0x01 0x00 0x00 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xa8` 

Operating Mode - {0x55, 0, 201, 1, -1, "..."},

- Fan: `0x55 0x00 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x66` 
- Dry: `0x55 0x07 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x5f` 
- Cool: `0x55 0x02 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x64` 
- Heat: `0x55 0x01 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x65` 
- Auto+Cool: `0x55 0x06 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x60 `
- Auto+Heat: `0x55 0x05 0x00 0x00 0x00 0x00 0x41 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x61 `

Target Temp - {0x54, 8, 153, 2, -1, "..."},

16c: `0x54 0x60 0x18 0x54 0x0e 0x73 0x19 0x00 0x00 0x00 0x10 0x00 0x00 0x00 0x00 0x00 0x00 0x35` 
17c: `0x54 0x48 0x18 0xf0 0x0f 0x73 0x19 0x00 0x00 0x00 0x11 0x00 0x00 0x00 0x00 0x00 0x00 0xaf` 
18c: `0x54 0x48 0x18 0x5c 0x10 0x73 0x19 0x00 0x00 0x00 0x12 0x00 0x00 0x00 0x00 0x00 0x00 0x41` 
19c: `0x54 0x74 0x18 0x0c 0x11 0x73 0x19 0x00 0x00 0x00 0x13 0x00 0x00 0x00 0x00 0x00 0x00 0x63` 
20c: `0x54 0x74 0x18 0xc8 0x10 0x5b 0x19 0x00 0x00 0x00 0x14 0x00 0x00 0x00 0x00 0x00 0x00 0xbf` 
21c: `0x54 0x48 0x18 0x88 0x10 0x5b 0x19 0x00 0x00 0x00 0x15 0x00 0x00 0x00 0x00 0x00 0x00 0x2a` 
...
32c: `0x54 0x74 0x18 0x88 0x10 0x5b 0x19 0x00 0x00 0x00 0x20 0x00 0x00 0x00 0x00 0x00 0x00 0xf3`

**Reply by revk** [reply]
Do we have a good name for this protocol? I'll have to add it, even if for reading.

Looks pretty simple.

**Reply by howlzj** [reply]
To avoid confusion, I propose we call this Altherma_S given these units seem to be using that platform?

Amazingly it's also called 'S' protocol in Daikin diagnostic software for Altherma units (https://github.com/raomin/ESPAltherma/issues/74#issuecomment-1039717140).

**Reply by revk** [reply]
OK, I may abbreviate for a setting name :-)

When I get on to this, ideally, I'll need you to point your MQTT to my server and allow me to load some test code, etc.

I'll let you know. Not sure exactly when I have time.

**Reply by revk** [reply]
I have a feeling this may be next week now, sorry, I'll see if I can fit in sooner though.

**Comment by revk** [comment]
I have made a beta with a setting `nosa` (default set), and when unset it will try Altherma_S protocol.

I have yet to code any of the messages.

When you have time, set your `mqtthost` to `testmqtt.revk.uk` and let me know when I have have a play with it (updating code, etc).

**Reply by howlzj** [reply]
Hi - done. Do you have a user/pass on your mqtt?

**Reply by revk** [reply]
Not for that test one.

**Reply by revk** [reply]
Good, I see it `state/faikin/status {"protocol":"Altherma_S","online":false,"autor":0.0,"autot":0.0,"auto0":"00:00","auto1":"00:00","autop":false}`

Today may be busy, but I'll see if I can make a start on this some time soon.

**Reply by revk** [reply]
Ah, off line now, I'll look later or in the morning, thanks.

**Reply by howlzj** [reply]
Apologies, my bad. Should be working now?

**Reply by revk** [reply]
Yes, I see it now.

OK I'll see if I can get on to this some time this week.

**Reply by revk** [reply]
Sorry for the delay, I am taking to it. Do you know if anyone has managed to make it control settings at all?

Anyway, now I have messages, extracting information should be pretty simple to add. I only see `U`, `T` and `S` getting replies at present.

What data would you like me to start with for testing?

```
info/faikin/tx {"protocol":"Altherma_S","cmd":"U","dump":"0255A8"}
info/faikin/rx {"protocol":"Altherma_S","cmd":"U","dump":"55070000000041030000000000000000005F"}
info/faikin/tx {"protocol":"Altherma_S","cmd":"T","dump":"0254A9"}
info/faikin/rx {"protocol":"Altherma_S","cmd":"T","dump":"5460175017D71700000018000000000000C7"}
info/faikin/tx {"protocol":"Altherma_S","cmd":"S","dump":"0253AA"}
info/faikin/rx {"protocol":"Altherma_S","cmd":"S","dump":"5300000001000000000000000000000000AB"}
```

**Reply by alexcf** [reply]
@howlzj did you have any luck getting this working?

**Reply by sveba** [reply]
@revk @howlzj I think I have the same protocol. Can I support somehow?

**Reply by Jamiga70** [reply]
creo que tengo el mismo protocolo. ¿Puedo apoyar tambien de alguna manera?

**Reply by howlzj** [reply]
> @howlzj did you have any luck getting this working?

Sorry - I have been quite distracted with personal matters for the last few months.

Unfortunately no - this seems to be an unsolved problem for [ESPAltherma](https://github.com/raomin/ESPAltherma/issues/1) as well. Their approach was to spoof a hardwired controller as an 'easier' option.

**Reply by Jamiga70** [reply]
Buenas, con los datos que recibo en RX "dump" ya puedo saber el estado de la máquina , temperatura de consigna, on, off, frio, calor y velocidad del ventilador, lo que no se, es si puedo enviar algun dato por TX para on , off.

**Comment by sveba** [comment]
I have a FTX25J3V1B. This is my Board. Connected to S403-port 
![PXL_20240525_073408793](https://github.com/revk/ESP32-Faikin/assets/1735151/3e6512dd-db00-4aa6-8d51-e6b249b702b5)

**Comment by bpetro1262** [comment]
@revk did you end up solving this one?  I have a fdyq250lav1 indoor unit (with X50A/X35 ports), and connected it using one of these cables: https://www.aliexpress.com/item/1005005465304996.html?spm=a2g0o.order_list.order_list_main.11.4e781802joU7C5

The Faikin powers on (purple light flashing) and I can connect to it via WiFi - the UI simply says "System is offline."

Any help / insights?

**Reply by alexcf** [reply]
There isn't support for our units just yet. I have the same one, but requires a bit of a deep dive. I have Madokas on mine, so I'm using an ESP32+BTLE solution for now.

**Reply by bpetro1262** [reply]
Thanks! Any reason why you chose Madoka vs Airbase BRP15B61?

**Reply by adrianoapmartins** [reply]
> I'm using an ESP32+BTLE solution for now.

@alexcf, could you elaborate on this? I’m in the same boat.

**Comment by MrSebBird** [comment]
It seems i am in the same boat when it comes to this aswell

My unit is a  FBQ60BV1A with a FBQ35-60B7V1 control board - This is an in ceiling unit and it currently is controled with a P1|P2 interface.

Like yours the X50A port is a larger JST-EH.

Happy to help put where i can in reverse engineering this.

Has there been any more progress?

My board below.

![Uploading IMG_0924.jpeg…]()

**Reply by adrianoapmartins** [reply]
Same boat as well. Also did some manual wiring but I also get the “System is offline" message.

**Comment by revk** [comment]
I can get some dumps from my X50A maybe.

**Reply by MrSebBird** [reply]
That will be super helpful

**Reply by MrSebBird** [reply]
how about i point my MQTT to you? Adding to that i have just changed protocol Altherma and i am seeing data as you did before. So it seems we are communicating.

**Reply by revk** [reply]
I have a server set up. I have some "real work" today as well, so I'll see what I can do.

**Comment by revk** [comment]
This is an example from my unit, X50A
```
2024-09-10 09:28:55.341240:info/OfficeAC/rx {"protocol":"X50A","dump":"06BE0F0106010241040101000001DA"}
2024-09-10 09:28:55.450194:info/OfficeAC/tx {"protocol":"X50A","dump":"06CA170100000000000000000000000000000000000017"}
2024-09-10 09:28:55.909416:info/OfficeAC/rx {"protocol":"X50A","dump":"06CA1701060102021102001100000000000000000200E6"}
2024-09-10 09:28:56.048686:info/OfficeAC/tx {"protocol":"X50A","dump":"06CB080100000025"}
2024-09-10 09:28:56.443005:info/OfficeAC/rx {"protocol":"X50A","dump":"06CB08010602110C"}
2024-09-10 09:28:57.245364:info/OfficeAC/tx {"protocol":"X50A","dump":"06BD06010035"}
2024-09-10 09:28:57.803561:info/OfficeAC/rx {"protocol":"X50A","dump":"06BD230106EA0BD40BF606EA0B9908EA0B0000000000005E120000E813D629046400E5"}
2024-09-10 09:28:57.810343:info/OfficeAC/tx {"protocol":"X50A","dump":"06BE06010034"}
2024-09-10 09:28:58.346323:info/OfficeAC/rx {"protocol":"X50A","dump":"06BE0F0106010242040101000001D9"}
2024-09-10 09:28:58.550723:info/OfficeAC/tx {"protocol":"X50A","dump":"06CA170100000000000000000000000000000000000017"}
2024-09-10 09:28:58.909607:info/OfficeAC/rx {"protocol":"X50A","dump":"06CA1701060102021102001100000000000000000200E6"}
2024-09-10 09:28:59.047798:info/OfficeAC/tx {"protocol":"X50A","dump":"06CB080100000025"}
2024-09-10 09:28:59.447884:info/OfficeAC/rx {"protocol":"X50A","dump":"06CB08010602110C"}
2024-09-10 09:29:00.242828:info/OfficeAC/tx {"protocol":"X50A","dump":"06BD06010035"}
2024-09-10 09:29:00.820232:info/OfficeAC/rx {"protocol":"X50A","dump":"06BD230106EA0BD40B1407EA0B9908F60B0000000000005E120000E813D629046400BA"}
2024-09-10 09:29:00.948351:info/OfficeAC/tx {"protocol":"X50A","dump":"06BE06010034"}
2024-09-10 09:29:01.341050:info/OfficeAC/rx {"protocol":"X50A","dump":"06BE0F0106010242040101000001D9"}
2024-09-10 09:29:01.448886:info/OfficeAC/tx {"protocol":"X50A","dump":"06CA170100000000000000000000000000000000000017"}
2024-09-10 09:29:01.909708:info/OfficeAC/rx {"protocol":"X50A","dump":"06CA1701060102021102001100000000000000000200E6"}
2024-09-10 09:29:02.048150:info/OfficeAC/tx {"protocol":"X50A","dump":"06CB080100000025"}
2024-09-10 09:29:02.457965:info/OfficeAC/rx {"protocol":"X50A","dump":"06CB08010602110C"}
2024-09-10 09:29:03.247803:info/OfficeAC/tx {"protocol":"X50A","dump":"06BD06010035"}
```

**Comment by bpetro1262** [comment]
I've decided to sell my Faikin unit since it doesn't work with my air con unit, and there doesn't appear to be progress on resolving it.  PM me if you want to buy it off me (I'm in Australia).

**Comment by alexcf** [comment]
Sure - I'm running this - https://github.com/Petapton/esphome/tree/madoka

**Comment by bryc3dev** [comment]
@revk @alexcf was there any progress on this? I also have a FDYQ unit with the Blue X50 plug

**Reply by alexcf** [reply]
Honestly I haven't bothered looking. Madoka and the ESP32 BTLE route is "ok" for now. I'd love to get this working properly, more of a time problem.
